sycl/include/sycl/ext/oneapi/experimental/group_load_store.hpp

```diff
@@ -106,6 +108,116 @@ int get_mem_idx(GroupTy g, int vec_or_array_idx) {
+
+  static constexpr int block_size =
+      sizeof(value_type) * (blocked ? ElementsPerWorkItem : 1);
+  static constexpr int num_blocks = blocked ? 1 : ElementsPerWorkItem;
+  static constexpr bool has_builtin =
+      detail::is_power_of_two(block_size) &&
+      detail::is_power_of_two(num_blocks) && block_size <= 8 &&
+      (num_blocks <= 8 || (num_blocks == 16 && block_size == 1));
+};
+
+template <typename BlockInfoTy> struct BlockTypeInfo;
+
+template <typename IteratorT, std::size_t ElementsPerWorkItem, bool blocked>
+struct BlockTypeInfo<BlockInfo<IteratorT, ElementsPerWorkItem, blocked>> {
+  using BlockInfoTy = BlockInfo<IteratorT, ElementsPerWorkItem, blocked>;
+  static_assert(BlockInfoTy::has_builtin);
+
+  using block_type = detail::cl_unsigned<BlockInfoTy::block_size>;
+
+  using block_pointer_elem_type = std::conditional_t<
+      std::is_const_v<std::remove_reference_t<
+          typename std::iterator_traits<IteratorT>::reference>>,
+      std::add_const_t<block_type>, block_type>;
+
+  using block_pointer_type = typename detail::DecoratedType<
+      block_pointer_elem_type, access::address_space::global_space>::type *;
+  using block_load_type = std::conditional_t<
+      BlockInfoTy::num_blocks == 1, block_type,
+      detail::ConvertToOpenCLType_t<vec<block_type, BlockInfoTy::num_blocks>>>;
+};
+
+// Returns either a pointer suitable to use in a block read/write builtin or
+// nullptr if some legality conditions aren't satisfied.
+template <int RequiredAlign, std::size_t ElementsPerWorkItem,
+          typename IteratorT, typename Properties>
+auto get_block_op_ptr(IteratorT iter, [[maybe_unused]] Properties props) {
+  using value_type =
+      remove_decoration_t<typename std::iterator_traits<IteratorT>::value_type>;
+  using iter_no_cv = std::remove_cv_t<IteratorT>;
+
+  constexpr bool blocked = detail::isBlocked(props);
+  using BlkInfo = BlockInfo<IteratorT, ElementsPerWorkItem, blocked>;
+
+#if defined(__SPIR__)
+  // TODO: What about non-Intel SPIR-V devices?
+  constexpr bool is_spir = true;
+#else
+  constexpr bool is_spir = false;
+#endif
+
+  if constexpr (!is_spir || !BlkInfo::has_builtin) {
+    return nullptr;
+  } else if constexpr (!props.template has_property<full_group_key>()) {
+    return nullptr;
+  } else if constexpr (detail::is_multi_ptr_v<IteratorT>) {
+    return get_block_op_ptr<RequiredAlign, ElementsPerWorkItem>(
+        iter.get_decorated(), props);
+  } else if constexpr (!std::is_pointer_v<iter_no_cv>) {
+    if constexpr (props.template has_property<contiguous_memory_key>())
+      return get_block_op_ptr<RequiredAlign, ElementsPerWorkItem>(&*iter,
+                                                                  props);
+    else
+      return nullptr;
+  } else {
+    __builtin_assume(iter != nullptr);
+    static_assert(BlkInfo::has_builtin);
+    bool aligned = alignof(value_type) >= RequiredAlign ||
+                   reinterpret_cast<uintptr_t>(iter) % RequiredAlign == 0;
+
```

Comment:
I see that when `value_type` is not aligned (i.e. `aligned` is false), we always return a `nullptr`. How about we do something like:

```
// Exit prematurely when value_type is not aligned.
if (!aligned) return nullptr;

constexpr auto AS = .....
```
This would simplify the latter code and we can even inline the `aligned` variable, as it will only be used once.

Reply:
We can't because it's not `constexpr` and breaks type deduction by creating ambiguity.

Reply:
Added a comment as well.